pages/api/recent-posts.ts

```diff
@@ -0,0 +1,110 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+
+import { ExtendedRecordMap } from 'notion-types'
+import {
+  getBlockParentPage,
+  getBlockTitle,
+  getPageProperty,
+  idToUuid
+} from 'notion-utils'
+
+import * as config from '@/lib/config'
+import { getSiteMap } from '@/lib/get-site-map'
+import { getSocialImageUrl } from '@/lib/get-social-image-url'
+import { getCanonicalPageUrl } from '@/lib/map-page-url'
+
+type BlogPost = {
+  title: string
+  description: string
+  published: number // unix timestamp
+  url: string
+  authors: string[]
+  coverImage: {
+    url: string
+    width: number
+    height: number
+  }
+}
+
+type Response = {
+  posts: BlogPost[]
+}
+
+/**
+ * This handler exposes recent blog posts as a JSON object for displaying post
+ * previews on the rest of the site.
+ */
+export default async function (req: NextApiRequest, res: NextApiResponse) {
+  const siteMap = await getSiteMap()
+  const ttl = 24 * 60 * 60 // 24 hours
+  const data: Response = {
+    posts: []
+  }
+
+  const limit = 5
+
+  for (const pagePath of Object.keys(siteMap.canonicalPageMap)) {
+    const pageId = siteMap.canonicalPageMap[pagePath]
+    const recordMap = siteMap.pageMap[pageId] as ExtendedRecordMap
+    if (!recordMap) continue
+
+    const keys = Object.keys(recordMap?.block || {})
+    const block = recordMap?.block?.[keys[0]]?.value
+    if (!block) continue
+
+    const parentPage = getBlockParentPage(block, recordMap)
+    const isBlogPost =
+      block.type === 'page' &&
+      block.parent_table === 'collection' &&
+      parentPage?.id === idToUuid(config.rootNotionPageId)
+    if (!isBlogPost) {
+      continue
+    }
+
+    const isPublished = getPageProperty<number>('Published', block, recordMap)
+    if (!isPublished) {
+      continue
+    }
+
+    const title = getBlockTitle(block, recordMap) || config.name
+    const description =
+      getPageProperty<string>('Description', block, recordMap) ||
+      config.description
+    const url = getCanonicalPageUrl(config.site, recordMap)(pageId)
+    const socialImageUrl = getSocialImageUrl(pageId)
+    const authorList =
+      getPageProperty<string>('Authors', block, recordMap) || config.author
+    const published = getPageProperty<number>('Date', block, recordMap)
+
+    // By default, social image URLs have a big title in front. We add an extra
+    // query parameter to signal that we want to use the image as-is.
+    const revisedImageUrl = new URL(socialImageUrl)
+    revisedImageUrl.searchParams.set('text', 'false')
+    const coverImageUrl = revisedImageUrl.toString()
```

Comment:
I don't think we're actually getting the right image url. In most cases it's blank? This is an existing issue. I'll see if there's a workaround.

Reply:
Nice! I noticed it was just some of the newer posts that had them, but if you're able to get it working for more, that's great. We're just pulling the latest few posts anyways, so it didn't seem like an issue.